### DISCUSSION TITLE
Testes Unitários em Aplicações JS

### DISCUSSION THREAD

**Discussion by mksanti** [frontendbr/forum#63] [Perguntas e Respostas]
que ferramentas estão utilizando para realizar testes ?

**Comment by victormiguez** [comment]
@mikejavier a issue #43 está tratando sobre ferramentas para teste de performance.
Não sei sobre qual tipo de testes você está se referindo, pode ser que as respostas dessa issue te ajude.

**Comment by vitorbritto** [comment]
Isso. Acompanha lá @mikejavier. Existem muitos tipos de testes para Software (em Front-End não é diferente). Caso a _issue_ seja para um teste específico (ou diferente da _issue_ #43), reabrimos ela. Combinado assim? =]

**Comment by mksanti** [comment]
Foi mal, acabei esquecendo de especificar. Estava me referindo a TDD e as ferramentas/frameworks utilizados para teste unitários, integração, etc, em app JS.

**Comment by vitorbritto** [comment]
Ok @mikejavier. Reabri a _issue_, mas editei o título pois TDD refere-se a uma metodologia para realização de testes em Softwares. Além disso, para mantermos o foco da discussão em um modelo de teste, coloquei o unitário. Combinado?

Eu vou abrir outra _issue_ sobre **Testes de Integração**, então, provavelmente, teremos referências para ferramentas também. =]

**Comment by mksanti** [comment]
show @vitorbritto. 
Quanto mais abordagens sobre o assunto melhor.

**Comment by doug2k1** [comment]
Aqui a gente usa Karma para executar e Jasmine para escrever os testes. Dá pra fazer TDD de uma forma bem legal. Vc coloca o Karma pra ficar monitorando os arquivos, e um plugin de notificação. Aí vc escreve o teste quebrado, sobe a notificação de falha (red), corrige a implementação, sobe a notificação de sucesso (green), refatora o código (refactor), sobe mais um sucesso, e assim vai. Dá pra fazer tudo sem tirar o foco do código.

A gente também usa outras ferramentas auxiliares (Sinon pra mockar o backend, jasmine-jquery pra facilitar testes de views, jsFactories pra construir dados de testes). Fiz essa apresentação mostrando um pouco desse workflow: http://www.slideshare.net/DouglasMatoso/testes-unitrios-de-js-com-jasmine-e-karma

**Comment by mksanti** [comment]
Booooaaaa @doug2k1, darei uma olhada nesses slides. Nesta assunto de teste sou muito leigo, vai para começar a ter um norte.

**Comment by ptcmariano** [comment]
Aqui a gente usa mocha. Antes usavamos o qunit, mas o código ficava muito verboso e com a alteração para mocha os testes ficaram mais legíveis.
Slides que explica um pouco: [http://slides.com/luizkota/testes-de-unidade-javascript-com-mocha-chai-e-istanbul](http://slides.com/luizkota/testes-de-unidade-javascript-com-mocha-chai-e-istanbul)

**Comment by thiamsantos** [comment]
Ressuscitando a issue, eu tenho usado ultimamente o [tape](https://github.com/substack/tape) por prover uma API simples e fácil de usar. Cheguei a usar o [ava](https://github.com/avajs/ava), mas o que incomodava era a lentidão.

Ultimamente tenho visto bastante gente falar do [Jest](http://facebook.github.io/jest/). Alguém já usa ou testou? Quais são as conclusões?

**Comment by bukinoshita** [comment]
Os que eu uso são:

- [AVA](https://github.com/avajs/ava) Nunca tive problema de lentidão como o @thiamsantos citou.
- [Enzyme](https://github.com/airbnb/enzyme)
- [Jest](https://github.com/facebook/jest)

**Comment by diessica** [comment]
Tive problemas de lentidão com o ava também. Achei que seria uma boa experiência num projeto grande, mas não foi. =/ Gosto da simplicidade dele, então só optaria por ele em coisas pequenas.

No momento, prefiro optar pelo clássico Mocha, ou Jest. Minha opção pelo Jest é mais recente, em razão das últimas atualizações dele que trouxeram muitas melhorias bacanas, de verdade.

> Ultimamente tenho visto bastante gente falar do Jest. 

Sim, pois o Jest melhorou muito! Está realmente muito legal e funciona sem você configurar **nada**. Só colocar no projeto e começar a testar! 

E se você precisar de algo mais específico, a API é completa e bem documentada (https://facebook.github.io/jest/docs/configuration.html).

Até o momento, não me faltou nada. E olha que precisei fazer várias coisas loucas como resolver módulos de forma diferente em alguns testes específicos. Recomendo!

**Comment by alanhr** [comment]
Eu gosto muito de usar o Mocha + Chai, sempre uso  essa combinação nos meus projetos
